File: 1297-maximum-number-of-balloons/maximum-number-of-balloons.cpp
class Solution {
public:
    int maxNumberOfBalloons(string text) {
        // // by the help of vector
        // // 128 size ka vector banaya (Saare symbols ke liye pehle se 0 initialized)
        // vector<int> freq(128, 0);
        
        // // Pass 1: Frequency count karna
        // for (char c : text) {
        //     freq[c]++;
        // }
        
        // // Pass 2: Har character ke hisab se balloon count nikalna
        // int b_count = freq['b'];
        // int a_count = freq['a'];
        // int l_count = freq['l'] / 2; // 2 'l' milkar 1 balloon banta hai
        // int o_count = freq['o'] / 2; // 2 'o' milkar 1 balloon banta hai
        // int n_count = freq['n'];
        
        // // C++ unique feature: curly braces ke sath ek baar mein sabka minimum
        // return min({b_count, a_count, l_count, o_count, n_count});

        // by the help of hashmap
        
        // Vector ko hata kar asli HashMap banaya
        unordered_map<char, int> freq;
        
        // Pass 1: Map mein characters ki frequency dynamic bharna
        for (char c : text) {
            freq[c]++; 
        }
        
        // Pass 2: Map se values nikalna
        int b_count = freq['b'];
        int a_count = freq['a'];
        int l_count = freq['l'] / 2;
        int o_count = freq['o'] / 2;
        int n_count = freq['n'];
        
        return min({b_count, a_count, l_count, o_count, n_count});
    }
};

File: 1297-maximum-number-of-balloons/maximum-number-of-balloons.java
class Solution {
    public int maxNumberOfBalloons(String text) {
        // 128 size ka primitive array (Pehle se 0 initialized)
        int[] freq = new int[128];
        
        // Pass 1: Frequency count karna
        for (int i = 0; i < text.length(); i++) {
            freq[text.charAt(i)]++;
        }
        
        // Pass 2: Balloon counts nikalna
        int b = freq['b'];
        int a = freq['a'];
        int l = freq['l'] / 2;
        int o = freq['o'] / 2;
        int n = freq['n'];
        
        // Java ka rule: Nested tarike se do-do karke minimum nikalna
        return Math.min(b, Math.min(a, Math.min(l, Math.min(o, n))));

        // // hashmap
        
        // // Array ko hata kar asli Java HashMap banaya
        // HashMap<Character, Integer> freq = new HashMap<>();
        
        // // Pass 1: Frequency count dynamic update karna
        // for (int i = 0; i < text.length(); i++) {
        //     char c = text.charAt(i);
        //     freq.put(c, freq.getOrDefault(c, 0) + 1);
        // }
        
        // // Pass 2: Map se values nikalna (Nahi milne par default 0 milega)
        // int b = freq.getOrDefault('b', 0);
        // int a = freq.getOrDefault('a', 0);
        // int l = freq.getOrDefault('l', 0) / 2;
        // int o = freq.getOrDefault('o', 0) / 2;
        // int n = freq.getOrDefault('n', 0);
        
        // return Math.min(b, Math.min(a, Math.min(l, Math.min(o, n))));
    }
}